src/runtime.js

```diff
@@ -469,15 +492,45 @@ function prepareAndRegisterElementShadowMode(global, extensions,
   addElementToExtension(extensions, name, implementationClass, opt_css);
   registerElementClass(global, name, implementationClass, opt_css);
   if (opt_css) {
-    addShadowRootFactoryToExtension(extensions, shadowRoot => {
-      installStylesForShadowRoot(shadowRoot, dev().assertString(opt_css),
-          /* isRuntimeCss */ false, name);
-    });
+    installExtensionStylesInShadow(extensions, name, opt_css);
   }
 }
 
 
 /**
+ * Registers an extended element and installs its styles in a shadow-doc mode
+ * for both `AmpDocShadow` and `AmpDocShell`
+ * @param {!Window} global
+ * @param {!./service/extensions-impl.Extensions} extensions
+ * @param {string} name
+ * @param {function(new:BaseElement)} implementationClass
+ * @param {string=} opt_css
+ */
+function prepareAndRegisterElementShadowModeWithAmpdocShell(global,
+    extensions, name, implementationClass, opt_css) {
+  prepareAndRegisterElement(global, extensions, name, implementationClass,
```

Comment:
FYI, I'm redoing this particular part right now in https://github.com/ampproject/amphtml/pull/10992

Let me know if you want to merge your code now (with the possibility of break) or if you want me to finish this migration first. Up to you.

Reply:
After reviewing #10992 this PR could actually be even simpler. Your changes in `runtime.js` can override the ones in this PR, and the only think that needs to be changed is here:

https://github.com/dvoytenko/amphtml/blob/dba27af195eb1ef2a027e7727db395a24b71445b/src/service/extensions-impl.js#L388

We should check if `isSingleDoc` or if it's `AmpDocShell`, so we install services or extensions right away for the `AmpDocShell`too. That's it!

I don't want you to get more conflicts than necessary in your PR, since it's quite massive, but if we land this PR first I can help you to fix them, I'm pretty sure it will be easy.